source/extensions/transformers/aws_lambda/api_gateway_transformer.cc

```diff
@@ -0,0 +1,202 @@
+#include "source/extensions/transformers/aws_lambda/api_gateway_transformer.h"
+
+#include "source/common/http/headers.h"
+#include "source/common/http/utility.h"
+#include "source/common/common/base64.h"
+
+#include "source/common/http/header_map_impl.h"
+
+#include "nlohmann/json.hpp"
+using json = nlohmann::json;
+
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AwsLambda {
+
+HttpFilters::Transformation::TransformerConstSharedPtr
+ApiGatewayTransformerFactory::createTransformer(
+    const Protobuf::Message &config,
+    Server::Configuration::CommonFactoryContext &context) {
+    MessageUtil::downcastAndValidate<const ApiGatewayTransformerProto &>(
+          config, context.messageValidationContext().staticValidationVisitor());
+  return std::make_shared<ApiGatewayTransformer>();
+}
+
+ApiGatewayTransformer::ApiGatewayTransformer(){}
+const Envoy::Http::LowerCaseString ApiGatewayTransformer::AMAZON_ERRORTYPE_HEADER = Envoy::Http::LowerCaseString("x-amzn-errortype");
+
+void ApiGatewayTransformer::format_error(
+  Http::ResponseHeaderMap &response_headers,
+  Buffer::Instance &body,
+  ApiGatewayError &error,
+  Http::StreamFilterCallbacks &stream_filter_callbacks) const {
+    ENVOY_STREAM_LOG(debug, "Returning error with message: " + std::string(error.message), stream_filter_callbacks);
+
+    // clear existing response headers
+    response_headers.clear();
+
+    response_headers.setStatus(error.status_code);
+    response_headers.setContentType("text/plain");
+
+    response_headers.setReferenceKey(AMAZON_ERRORTYPE_HEADER, error.code);
+    body.drain(body.length());
+    auto error_message = absl::StrCat(error.code, ": ", error.message);
+    body.add(error_message);
+    response_headers.setContentLength(body.length());
+  }
+
+void ApiGatewayTransformer::transform(
+    Http::RequestOrResponseHeaderMap &header_map,
+    Http::RequestHeaderMap *request_headers,
+    Buffer::Instance &body,
+    Http::StreamFilterCallbacks &stream_filter_callbacks) const {
+      ENVOY_STREAM_LOG(debug, "Transforming response", stream_filter_callbacks);
+      // check if request header map equals response header map
+      if (*request_headers == header_map) {
+        // we are on the request path and we don't want to be
+        ENVOY_STREAM_LOG(debug, "Api Gateway transformer cannot be used on the request path", stream_filter_callbacks);
+        return;
+      }
+      // we know we are on the repsonse path, so downcast RequestResponseHeaderMap to ResponseHeaderMap
+      Http::ResponseHeaderMap* response_headers = static_cast<Http::ResponseHeaderMap*>(&header_map);
+      try {
+        transform_response(response_headers, body, stream_filter_callbacks);
+      } catch (const std::exception &e) {
+        ENVOY_STREAM_LOG(debug, "Error transforming response: " + std::string(e.what()), stream_filter_callbacks);
+        ApiGatewayError error = ApiGatewayError{400, "400", "Failed to transform response"};
```

Comment:
this feels more like a 5xx error than 4xx. the request was not malformed, it was a server error transforming the response

Reply:
changed [d809aac](https://github.com/solo-io/envoy-gloo/pull/204/commits/d809aac6ac333bb6af35b37a38aa5969f559d89f)